service/policy/keymanagement/key_management.go

```diff
@@ -0,0 +1,173 @@
+
+	rsp.ProviderConfig = pc
+
+	return connect.NewResponse(rsp), nil
+}
+
+func (ksvc Service) GetProviderConfig(ctx context.Context, req *connect.Request[keyMgmtProto.GetProviderConfigRequest]) (*connect.Response[keyMgmtProto.GetProviderConfigResponse], error) {
+	rsp := &keyMgmtProto.GetProviderConfigResponse{}
+
+	switch req := req.Msg.GetIdentifier().(type) {
+	case *keyMgmtProto.GetProviderConfigRequest_Id:
+		ksvc.logger.Debug("Getting Provider config by ID", slog.String("ID", req.Id))
+	case *keyMgmtProto.GetProviderConfigRequest_Name:
+		ksvc.logger.Debug("Getting Provider config by Name", slog.String("Name", req.Name))
+	default:
+		return nil, connect.NewError(connect.CodeInvalidArgument, nil)
+	}
+
+	auditParams := audit.PolicyEventParams{
+		ActionType: audit.ActionTypeCreate,
+		ObjectType: audit.ObjectTypeKeyManagementProviderConfig,
+	}
+
+	pc, err := ksvc.dbClient.GetProviderConfig(ctx, req.Msg.GetIdentifier())
+	if err != nil {
+		ksvc.logger.Audit.PolicyCRUDFailure(ctx, auditParams)
+		return nil, db.StatusifyError(err, db.ErrTextGetRetrievalFailed, slog.String("keyManagementService", req.Msg.String()))
+	}
+
+	auditParams.ObjectID = pc.GetId()
+	auditParams.Original = pc
+	ksvc.logger.Audit.PolicyCRUDSuccess(ctx, auditParams)
+
+	rsp.ProviderConfig = pc
+
+	return connect.NewResponse(rsp), nil
+}
+
+func (ksvc Service) ListProviderConfigs(ctx context.Context, req *connect.Request[keyMgmtProto.ListProviderConfigsRequest]) (*connect.Response[keyMgmtProto.ListProviderConfigsResponse], error) {
+	ksvc.logger.Debug("Listing Provider Configs")
+
+	resp, err := ksvc.dbClient.ListProviderConfigs(ctx, req.Msg.GetPagination())
+	if err != nil {
+		return nil, db.StatusifyError(err, db.ErrTextGetRetrievalFailed, slog.String("keyManagementService", req.Msg.String()))
+	}
+
+	return connect.NewResponse(resp), nil
+}
+
+func (ksvc Service) UpdateProviderConfig(ctx context.Context, req *connect.Request[keyMgmtProto.UpdateProviderConfigRequest]) (*connect.Response[keyMgmtProto.UpdateProviderConfigResponse], error) {
+	rsp := &keyMgmtProto.UpdateProviderConfigResponse{}
+	providerConfigID := req.Msg.GetId()
+
+	ksvc.logger.Debug("Updating Provider Config", slog.String("id", req.Msg.GetId()))
+
+	auditParams := audit.PolicyEventParams{
+		ActionType: audit.ActionTypeUpdate,
+		ObjectType: audit.ObjectTypeKeyManagementProviderConfig,
+		ObjectID:   providerConfigID,
+	}
+
+	original, err := ksvc.dbClient.GetProviderConfig(ctx, &keyMgmtProto.GetProviderConfigRequest_Id{
+		Id: providerConfigID,
+	})
+	if err != nil {
+		ksvc.logger.Audit.PolicyCRUDFailure(ctx, auditParams)
+		return nil, db.StatusifyError(err, db.ErrTextGetRetrievalFailed, slog.String("id", providerConfigID))
+	}
+
+	pc, err := ksvc.dbClient.UpdateProviderConfig(ctx, req.Msg)
+	if err != nil {
+		ksvc.logger.Audit.PolicyCRUDFailure(ctx, auditParams)
+		return nil, db.StatusifyError(err, db.ErrTextUpdateFailed, slog.String("keyManagementService", req.Msg.GetId()))
+	}
```

Comment:
I would advise using a transaction for this in case customers employ db replicas so you don't unintentionally introduce a data consistency problem. Here's an example: https://github.com/opentdf/platform/blob/main/service/policy/subjectmapping/subject_mapping.go#L86

Reply:
Totally accurate, but the Get above could be out of sync with the Update. There's no guarantee the service handler would hit the same db replica for both roundtrips. In this case, the only downside is a theoretical data inconsistency in audit logs, so I'm not too concerned about merging this as-is, but just something to think about.